sycl/include/sycl/ext/oneapi/experimental/uniform.hpp

```diff
@@ -0,0 +1,100 @@
+//==------ uniform.hpp - SYCL uniform extension --------*- C++ -*-----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+// Implemenation of the sycl_ext_oneapi_uniform extension.
+// https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_uniform.asciidoc
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+
+// 1 - Initial extension version. Base features are supported.
+#define SYCL_EXT_ONEAPI_UNIFORM 1
+
+#include <type_traits>
+
+namespace sycl {
+// Forward declarations of types not allowed to be wrapped in uniform:
+
+template <int, bool> class item;
+template <int> class nd_item;
+template <int> class h_item;
+template <int> class group;
+template <int> class nd_range;
+
+namespace ext {
+namespace oneapi {
+
+struct sub_group;
+
+namespace experimental {
+
+namespace detail {
+template <class T, template <int> class Tmpl>
+struct is_instance_of_tmpl_int : std::false_type {};
+template <int N, template <int> class T, template <int> class Tmpl>
+struct is_instance_of_tmpl_int<T<N>, Tmpl>
+    : std::conditional<std::is_same_v<T<N>, Tmpl<N>>, std::true_type,
+                       std::false_type> {};
+template <class T, template <int> class Tmpl>
+static inline constexpr bool is_instance_of_tmpl_int_v =
+    is_instance_of_tmpl_int<T, Tmpl>::value;
+
+template <class T, template <int, bool> class Tmpl>
+struct is_instance_of_tmpl_int_bool : std::false_type {};
+template <int N, bool X, template <int, bool> class T,
+          template <int, bool> class Tmpl>
+struct is_instance_of_tmpl_int_bool<T<N, X>, Tmpl>
+    : std::conditional<std::is_same_v<T<N, X>, Tmpl<N, X>>, std::true_type,
+                       std::false_type> {};
+template <class T, template <int, bool> class Tmpl>
+static inline constexpr bool is_instance_of_tmpl_int_bool_v =
+    is_instance_of_tmpl_int_bool<T, Tmpl>::value;
+} // namespace detail
+
+template <class T> class uniform {
+  template <class U> static constexpr bool can_be_uniform() {
+    return !detail::is_instance_of_tmpl_int_v<U, sycl::id> &&
+           !detail::is_instance_of_tmpl_int_bool_v<U, sycl::item> &&
+           !detail::is_instance_of_tmpl_int_v<U, sycl::nd_item> &&
+           !detail::is_instance_of_tmpl_int_v<U, sycl::h_item> &&
+           !detail::is_instance_of_tmpl_int_v<U, sycl::group> &&
+           !detail::is_instance_of_tmpl_int_v<U, sycl::nd_range> &&
+           !std::is_same_v<U, sycl::ext::oneapi::sub_group>;
+  }
+  static_assert(can_be_uniform<T>() && "type not allowed to be `uniform`");
+
+public:
+  explicit uniform(T x) noexcept : Val(x) {}
+  operator const T() const { return Val; }
```

Comment:
Would you mind adding a TODO here saying that we need to add the logic that allows the SPMD compilation flow to recognize the result of this conversion as uniform?

Reply:
Do you mean cases like 'auto x = invoke_simd(sg, SIMD_CALLEE,...);`, where SIMD_CALLEE returns `uniform<T>` or something more general regardless of `invoke_simd`?

Reply:
The latter.  The extension document has [some examples](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_uniform.asciidoc#example-usage) of ways we'd like people to be able to use `uniform` as an optimization hint.  What you've implemented is enough for those examples to compile, but the compiler won't actually optimize anything until we add something like a built-in inside this conversion.

Reply:
ok, makes sense.